Incorrect Usage of `executeQuery` for Data Modification

### Mautic Series

5.1.x series

### Mautic installed version

5.1.1

### Way of installing

I downloaded a release from https://www.mautic.org/mautic-releases

### PHP version

8.1

### What browsers are you seeing the problem on?

Not relevant

### What happened?

It has been identified that the `executeQuery` method is being used for data modification in the Mautic source code. This method is intended solely for reading data from the database. Using it for data modification, especially in the context of database replication, can lead to operations being executed directly on the replica. This can cause synchronization errors, such as key duplication.

Example: https://github.com/mautic/mautic/blob/5.x/app/bundles/CampaignBundle/Entity/SummaryRepository.php#L139

### Expected Behavior:
The `executeQuery` method should only be used for data retrieval. Data modification operations should use `executeStatement`.

Additional context info from Doctrine:
```
/**
 * Primary-Replica Connection
 *
 * Connection can be used with primary-replica setups.
 *
 * Important for the understanding of this connection should be how and when
 * it picks the replica or primary.
 *
 * 1. Replica if primary was never picked before and ONLY if 'getWrappedConnection'
 *    or 'executeQuery' is used.
 * 2. Primary picked when 'executeStatement', 'insert', 'delete', 'update', 'createSavepoint',
 *    'releaseSavepoint', 'beginTransaction', 'rollback', 'commit' or 'prepare' is called.
 * 3. If Primary was picked once during the lifetime of the connection it will always get picked afterwards.
 * 4. One replica connection is randomly picked ONCE during a request.
 *
 * ATTENTION: You can write to the replica with this connection if you execute a write query without
 * opening up a transaction. For example:
 *
 *      $conn = DriverManager::getConnection(...);
 *      $conn->executeQuery("DELETE FROM table");
 *
 * Be aware that Connection#executeQuery is a method specifically for READ
 * operations only.
 *
 * Use Connection#executeStatement for any SQL statement that changes/updates
 * state in the database (UPDATE, INSERT, DELETE or DDL statements). 
```

### How can we reproduce this issue?

1. **Configure the Replica Database:**
   - Connect a replica database to your Mautic setup.
   - Add the replica database host to the `local.php` configuration file using the key `db_host_ro`.

2. **Create and Execute a Test Campaign:**
   - Set up a test campaign within Mautic.
   - Execute the campaign to generate data that will be written to the database.

3. **Verify Data Consistency:**
   - Check the `campaign_summary` table in both the primary database and the replica.
   - Ensure that the data in `campaign_summary` is consistent between the primary database and the replica.

**Note:** This is just one identified case of `executeQuery` being used for data modification. There may be other instances where `executeQuery` is improperly used for write operations.

### Relevant log output
```
[ERROR] Slave SQL: Could not execute Write_rows_v1 event on table mautic.campaign_summary; Duplicate entry '4725-50157-2024-10-18 13:00:00' for key 'campaign_event_date_triggered', Error_code: 1062; handler error HA_ERR_FOUND_DUPP_KEY; the event's master log mariadb-bin.000460, end_log_pos 827533312, Internal MariaDB error code: 1062
[Warning] Slave: Duplicate entry '4725-50157-2024-10-18 13:00:00' for key 'campaign_event_date_triggered' Error_code: 1062
[ERROR] Error running query, slave SQL thread aborted. Fix the problem, and restart the slave SQL thread with "SLAVE START". We stopped at log 'mariadb-bin.000460' position 827531200
[Note] Slave SQL thread exiting, replication stopped in log 'mariadb-bin.000460' at position 827531200
```
_No response_

### Code of Conduct

- [X] I confirm that I have read and agree to follow this project's Code of Conduct

<br /><hr>
Care about this issue? Want to get it resolved sooner? If you are a <a href='https://www.mautic.org/become-a-member-of-mautic'>member of Mautic</a>, you can add some funds to the <a href='https://opencollective.com/mautic/projects/bounties'>Bounties Project</a> so that the person who completes this task can claim those funds once it is merged by a member of the core team! Read the docs <a href='https://contribute.mautic.org/product-team/mautic-bounty-programme'>here.</a>

--- a/app/bundles/CampaignBundle/Entity/LeadEventLogRepository.php
+++ b/app/bundles/CampaignBundle/Entity/LeadEventLogRepository.php
@@ -625,7 +625,7 @@ SQL;
         $conn          = $this->getEntityManager()->getConnection();
         $deleteEntries = true;
         while ($deleteEntries) {
-            $deleteEntries = $conn->executeQuery($sql, [$campaignId], [Types::INTEGER])->rowCount();
+            $deleteEntries = $conn->executeStatement($sql, [$campaignId], [Types::INTEGER]);
         }
     }
 
@@ -639,7 +639,7 @@ SQL;
         $conn          = $this->getEntityManager()->getConnection();
         $deleteEntries = true;
         while ($deleteEntries) {
-            $deleteEntries = $conn->executeQuery($sql, [$eventIds], [ArrayParameterType::INTEGER])->rowCount();
+            $deleteEntries = $conn->executeStatement($sql, [$eventIds], [ArrayParameterType::INTEGER]);
         }
     }
 

--- a/app/bundles/CampaignBundle/Entity/SummaryRepository.php
+++ b/app/bundles/CampaignBundle/Entity/SummaryRepository.php
@@ -138,7 +138,7 @@ class SummaryRepository extends CommonRepository
             ' triggered_count = s.triggered_count_i, '.
             ' log_counts_processed = s.log_counts_processed_i;';
 
-            $this->getEntityManager()->getConnection()->executeQuery($sql);
+            $this->getEntityManager()->getConnection()->executeStatement($sql);
         }
     }
 }

--- a/app/bundles/CampaignBundle/Tests/Functional/Entity/LeadEventLogRepositoryTest.php
+++ b/app/bundles/CampaignBundle/Tests/Functional/Entity/LeadEventLogRepositoryTest.php
@@ -21,11 +21,11 @@ class LeadEventLogRepositoryTest extends MauticMysqlTestCase
 
         $insertStatement = $connection->prepare('INSERT INTO `'.MAUTIC_TABLE_PREFIX.'campaign_lead_event_log` (`event_id`, `lead_id`, `rotation`, `is_scheduled`, `system_triggered`) VALUES (?, ?, ?, ?, ?);');
 
-        $connection->executeQuery('SET FOREIGN_KEY_CHECKS=0;');
+        $connection->executeStatement('SET FOREIGN_KEY_CHECKS=0;');
         foreach ($this->getLeadCampaignEventData($eventId) as $row) {
-            $insertStatement->executeQuery($row);
+            $insertStatement->executeStatement($row);
         }
-        $connection->executeQuery('SET FOREIGN_KEY_CHECKS=1;');
+        $connection->executeStatement('SET FOREIGN_KEY_CHECKS=1;');
 
         Assert::assertCount(3, $leadEventLogRepository->findAll());
 

--- a/app/bundles/CoreBundle/Entity/AuditLogRepository.php
+++ b/app/bundles/CoreBundle/Entity/AuditLogRepository.php
@@ -261,6 +261,6 @@ class AuditLogRepository extends CommonRepository
         $sql        = "UPDATE {$table_name} SET ip_address = '*.*.*.*' WHERE ip_address != '*.*.*.*'";
         $conn       = $this->getEntityManager()->getConnection();
 
-        return $conn->executeQuery($sql)->rowCount();
+        return $conn->executeStatement($sql);
     }
 }

--- a/app/bundles/CoreBundle/Entity/IpAddressRepository.php
+++ b/app/bundles/CoreBundle/Entity/IpAddressRepository.php
@@ -114,6 +114,6 @@ SQL;
         $sql               = "UPDATE {$table_name} SET ip_address = '*.*.*.*', ip_details = 'N;' WHERE ip_address != '*.*.*.*'";
         $conn              = $this->getEntityManager()->getConnection();
 
-        return $conn->executeQuery($sql)->rowCount();
+        return $conn->executeStatement($sql);
     }
 }

--- a/app/bundles/CoreBundle/EventListener/MigrationCommandSubscriber.php
+++ b/app/bundles/CoreBundle/EventListener/MigrationCommandSubscriber.php
@@ -60,7 +60,7 @@ class MigrationCommandSubscriber implements EventSubscriberInterface
             $output->writeln("<info>++</info> adding generated column <comment>{$generatedColumn->getColumnName()}</comment>");
             $output->writeln("<comment>-></comment> {$generatedColumn->getAlterTableSql()}");
 
-            $this->connection->executeQuery($generatedColumn->getAlterTableSql());
+            $this->connection->executeStatement($generatedColumn->getAlterTableSql());
 
             $duration = (string) $stopwatch->stop($generatedColumn->getColumnName());
             $output->writeln("<info>++</info> generated column added ({$duration})");

--- a/app/bundles/CoreBundle/EventListener/OptimisticLockSubscriber.php
+++ b/app/bundles/CoreBundle/EventListener/OptimisticLockSubscriber.php
@@ -54,7 +54,7 @@ class OptimisticLockSubscriber implements EventSubscriber
                 return "{$name} = :{$name}";
             }, $metadata->getIdentifierFieldNames())))
             ->setParameters($entityManager->getUnitOfWork()->getEntityIdentifier($object))
-            ->executeQuery();
+            ->executeStatement();
 
         $newVersion = (int) $connection->executeQuery('SELECT @newVersion')->fetchOne();
         $object->setVersion($newVersion);

--- a/app/bundles/CoreBundle/Test/MauticMysqlTestCase.php
+++ b/app/bundles/CoreBundle/Test/MauticMysqlTestCase.php
@@ -145,7 +145,7 @@ abstract class MauticMysqlTestCase extends AbstractMauticTestCase
         $connection = $this->connection;
 
         foreach ($tables as $table) {
-            $connection->executeQuery(sprintf('ALTER TABLE `%s%s` AUTO_INCREMENT=1', $prefix, $table));
+            $connection->executeStatement(sprintf('ALTER TABLE `%s%s` AUTO_INCREMENT=1', $prefix, $table));
         }
     }
 
@@ -160,7 +160,7 @@ abstract class MauticMysqlTestCase extends AbstractMauticTestCase
         $prefix = MAUTIC_TABLE_PREFIX;
 
         foreach ($tables as $table) {
-            $this->connection->executeQuery("SET FOREIGN_KEY_CHECKS = 0; TRUNCATE TABLE `{$prefix}{$table}`; SET FOREIGN_KEY_CHECKS = 1;");
+            $this->connection->executeStatement("SET FOREIGN_KEY_CHECKS = 0; TRUNCATE TABLE `{$prefix}{$table}`; SET FOREIGN_KEY_CHECKS = 1;");
         }
     }
 

--- a/app/bundles/CoreBundle/Tests/Unit/EventListener/MigrationCommandSubscriberTest.php
+++ b/app/bundles/CoreBundle/Tests/Unit/EventListener/MigrationCommandSubscriberTest.php
@@ -154,7 +154,7 @@ class MigrationCommandSubscriberTest extends \PHPUnit\Framework\TestCase
             ->willReturn(['id' => new \stdClass()]);
 
         $this->connection->expects($this->once())
-            ->method('executeQuery');
+            ->method('executeStatement');
 
         $this->subscriber->addGeneratedColumns($this->event);
     }

--- a/app/bundles/FormBundle/Tests/Controller/SubmissionFunctionalTest.php
+++ b/app/bundles/FormBundle/Tests/Controller/SubmissionFunctionalTest.php
@@ -507,7 +507,7 @@ final class SubmissionFunctionalTest extends MauticMysqlTestCase
         $tablePrefix = static::getContainer()->getParameter('mautic.db_table_prefix');
 
         if ($this->connection->createSchemaManager()->tablesExist("{$tablePrefix}form_results_1_submission")) {
-            $this->connection->executeQuery("DROP TABLE {$tablePrefix}form_results_1_submission");
+            $this->connection->executeStatement("DROP TABLE {$tablePrefix}form_results_1_submission");
         }
     }
 

--- a/app/bundles/FormBundle/Tests/EventListener/CampaignSubscriberFunctionalTest.php
+++ b/app/bundles/FormBundle/Tests/EventListener/CampaignSubscriberFunctionalTest.php
@@ -160,7 +160,7 @@ class CampaignSubscriberFunctionalTest extends MauticMysqlTestCase
         $tablePrefix = static::getContainer()->getParameter('mautic.db_table_prefix');
 
         if ($this->connection->createSchemaManager()->tablesExist("{$tablePrefix}form_results_1_test_form")) {
-            $this->connection->executeQuery("DROP TABLE {$tablePrefix}form_results_1_test_form");
+            $this->connection->executeStatement("DROP TABLE {$tablePrefix}form_results_1_test_form");
         }
     }
 }

--- a/app/bundles/InstallBundle/Helper/SchemaHelper.php
+++ b/app/bundles/InstallBundle/Helper/SchemaHelper.php
@@ -164,7 +164,7 @@ class SchemaHelper
         // Execute drop queries
         foreach ($sql as $q) {
             try {
-                $this->db->executeQuery($q);
+                $this->db->executeStatement($q);
             } catch (\Exception $exception) {
                 $this->db->close();
 

--- a/app/bundles/InstallBundle/Tests/Install/InstallSchemaTest.php
+++ b/app/bundles/InstallBundle/Tests/Install/InstallSchemaTest.php
@@ -101,7 +101,7 @@ class InstallSchemaTest extends TestCase
         if (!empty($sql)) {
             foreach ($sql as $q) {
                 try {
-                    $this->connection->executeQuery($q);
+                    $this->connection->executeStatement($q);
                 } catch (\Exception $exception) {
                     $exceptions[] = $exception->getMessage();
                 }

--- a/app/bundles/LeadBundle/Entity/CompanyLeadRepository.php
+++ b/app/bundles/LeadBundle/Entity/CompanyLeadRepository.php
@@ -217,7 +217,7 @@ class CompanyLeadRepository extends CommonRepository
         $conn = $this->getEntityManager()->getConnection();
         do {
             $sql = 'DELETE FROM '.MAUTIC_TABLE_PREFIX.'companies_leads WHERE is_primary = 0 LIMIT '.self::DELETE_BATCH_SIZE;
-            $row = $conn->executeQuery($sql)->rowCount();
+            $row = $conn->executeStatement($sql);
         } while ($row);
     }
 

--- a/app/bundles/PluginBundle/Bundle/PluginDatabase.php
+++ b/app/bundles/PluginBundle/Bundle/PluginDatabase.php
@@ -47,12 +47,12 @@ class PluginDatabase
             // Check if the query is a DDL statement
             if (self::isDDLStatement($q)) {
                 // Execute DDL statements outside of a transaction
-                $connection->executeQuery($q);
+                $connection->executeStatement($q);
             } else {
                 // For non-DDL statements, use transactions
                 try {
                     $connection->beginTransaction();
-                    $connection->executeQuery($q);
+                    $connection->executeStatement($q);
                     $connection->commit();
                 } catch (\Exception $e) {
                     // Rollback only for non-DDL statements
@@ -92,7 +92,7 @@ class PluginDatabase
         $db->beginTransaction();
         try {
             foreach ($dropQueries as $q) {
-                $db->executeQuery($q);
+                $db->executeStatement($q);
             }
 
             $db->commit();

--- a/app/bundles/WebhookBundle/Entity/LogRepository.php
+++ b/app/bundles/WebhookBundle/Entity/LogRepository.php
@@ -50,7 +50,7 @@ class LogRepository extends CommonRepository
 
         if ($id) {
             $sql = "DELETE FROM {$table_name} WHERE webhook_id = (?) and id <= (?) LIMIT ".self::LOG_DELETE_BATCH_SIZE;
-            while ($rows = $conn->executeQuery($sql, [$webHookId, $id], [ParameterType::INTEGER, ParameterType::INTEGER])->rowCount()) {
+            while ($rows = $conn->executeStatement($sql, [$webHookId, $id], [ParameterType::INTEGER, ParameterType::INTEGER])) {
                 $deletedLogs += $rows;
             }
         }

